hazelcast/src/main/java/com/hazelcast/client/impl/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -479,278 +682,205 @@ public void onError(Channel channel, Throwable cause) {
-            ClientConnection connection = activeConnections.get(inetSocketAddressCache.get(target));
-            if (connection != null) {
-                return connection;
-            }
-            AddressProvider addressProvider = currentClusterContext.getAddressProvider();
-            Address address = null;
-            try {
-                address = addressProvider.translate(target);
-            } catch (Exception e) {
-                logger.warning("Failed to translate address " + target + " via address provider " + e.getMessage());
-            }
-            if (address == null) {
-                throw new NullPointerException("Address Provider " + addressProvider.getClass()
-                        + " could not translate address " + target);
+                logger.info("Authenticated with server " + result.address + ", server version:" + connection
+                        .getConnectedServerVersion() + " Local address: " + connection.getLocalSocketAddress());
+
+                establishConnectedState(newClusterId, changedCluster);
+
+                // Check if connection is closed by remote before authentication complete, if that is the case
+                // we need to remove it back from active connections.
+                // Race description from https://github.com/hazelcast/hazelcast/pull/8832.(A little bit changed)
+                // - open a connection client -> member
+                // - send auth message
+                // - receive auth reply -> reply processing is offloaded to an executor. Did not start to run yet.
+                // - member closes the connection -> the connection is trying to removed from map
+                //                                                     but it was not there to begin with
+                // - the executor start processing the auth reply -> it put the connection to the connection map.
+                // - we end up with a closed connection in activeConnections map
+                if (!connection.isAlive()) {
+                    removeFromActiveConnections(connection);
+                }
+                break;
+            case CREDENTIALS_FAILED:
+                throw new AuthenticationException("Invalid credentials!");
+            case NOT_ALLOWED_IN_CLUSTER:
+                throw new ClientNotAllowedInClusterException("Client is not allowed in the cluster");
+            default:
+                throw new AuthenticationException("Authentication status code not supported. status: "
+                        + authenticationStatus);
+        }
+    }
+
+    private void establishConnectedState(UUID newClusterId, boolean changedCluster) {
+        boolean isFirstConnectionAfterDisconnection = connectionCount.incrementAndGet() == 1;
+        if (isFirstConnectionAfterDisconnection) {
+            if (changedCluster) {
+                clusterConnectionExecutor.submit(new SendStatesToCluster(newClusterId));
```

Comment:
I think it should not.  I choose to prevent other invocations via `checkInvocationAllowed` instead of waiting for this thread. Because waiting for this thread only possible in the initial start. In reconnect situation, other threads have no idea about reconnection. 

Moreover, sending a cluster state is not strictly related to opening this connection. If we make this part blocking, it will keep holding the `lock` to opening the connection. Secondly, a fail in sending the state should be handled and retried. I don't want this thread which is just trying to open a connection, to wait in sending the state to cluster. It will wait further in the line via looking at `checkInvocationAllowed`.